llvm/lib/Analysis/CFG.cpp

```diff
@@ -130,14 +130,35 @@ static const Loop *getOutermostLoop(const LoopInfo *LI, const BasicBlock *BB) {
   return L ? L->getOutermostLoop() : nullptr;
 }
 
-bool llvm::isPotentiallyReachableFromMany(
-    SmallVectorImpl<BasicBlock *> &Worklist, const BasicBlock *StopBB,
-    const SmallPtrSetImpl<BasicBlock *> *ExclusionSet, const DominatorTree *DT,
-    const LoopInfo *LI) {
-  // When the stop block is unreachable, it's dominated from everywhere,
+template <class T, bool IsMany>
+static bool isReachableImpl(SmallVectorImpl<BasicBlock *> &Worklist,
+                            const T *StopBBOrSet,
+                            const SmallPtrSetImpl<BasicBlock *> *ExclusionSet,
+                            const DominatorTree *DT, const LoopInfo *LI) {
+  const BasicBlock *StopBB;
+  const SmallPtrSetImpl<const BasicBlock *> *StopSet;
+
+  // SmallPtrSetImpl is incompatible with LLVM's casting functions.
+  if constexpr (IsMany)
+    StopSet =
+        static_cast<const SmallPtrSetImpl<const BasicBlock *> *>(StopBBOrSet);
+  else
+    StopBB = static_cast<const BasicBlock *>(StopBBOrSet);
```

Comment:
This is all pretty ugly ... could the code be simplified by using something like this maybe?
```
class SingleEntrySet {
  T Elem;
  SingleEntrySet(T Elem) : Elem(Elem) {}
  bool contains(T Other) const { return Elem == Other; }
}
```

Reply:
If I understand your comment correctly, you're suggesting that I pass a `SingleEntrySet<const BasicBlock *>` to `isReachableImpl(...)` when implementing `isPotentiallyReachableFromMany(...)` in order to remove all the `if` statements. This would require `SingleEntrySet` to be iterable (since we iterate through the elements of `StopSet`), which drives up the complexity of the class further.

(We could do something like

```cpp
template <class T>
class SingleEntrySet {
  using iterator = T*;
  T Elem;  
  iterator begin() { return &Elem; }
  iterator end() { return &Elem + 1; }
};
```

but this is UB)

Reply:
Friendly ping @nikic - any comments?

Reply:
> but this is UB

Why would that be UB?

Reply:
Something like https://github.com/nikic/llvm-project/commit/932822472eeba688e5bfcb6a3ffc6c0f79f69dd4, though it would be nice to also avoid the StopLoops conditional...

I don't think we need StopBBReachable -- setting DT to nullptr if any one is unreachable is fine here (DT is only an optimization, and unreachable block is an unlikely edge case).